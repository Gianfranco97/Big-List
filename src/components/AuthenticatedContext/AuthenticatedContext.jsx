import React, { useMemo, useState, createContext } from 'react';

const AuthenticatedContext = createContext({
  isAuthenticated: false,
  changeAuthenticatedStatus: () => {},
});

export function AuthenticatedProvider(props) {
  const [isAuthenticated, changeAuthenticatedStatus] = useState(
    !!sessionStorage.getItem('session-token'),
  );

  const value = useMemo(
    () => ({
      isAuthenticated,
      changeAuthenticatedStatus,
    }),
    [isAuthenticated],
  );

  return <AuthenticatedContext.Provider value={value} {...props} />;
}

export default AuthenticatedContext;
