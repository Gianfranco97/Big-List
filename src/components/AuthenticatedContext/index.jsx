export {
  default,
  AuthenticatedProvider,
} from './AuthenticatedContext';
