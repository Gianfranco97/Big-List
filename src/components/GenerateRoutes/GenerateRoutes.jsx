import React from 'react';
import PropTypes from 'prop-types';
import { Navigate, Route, Routes } from 'react-router-dom';
import MissingPage from '../MissingPage';
import AuthenticatedContext from '../AuthenticatedContext';

function RouteRender({ isPrivate, RenderComponent, isAuthenticated }) {
  return isPrivate && !isAuthenticated ? (
    <Navigate to="/login" />
  ) : (
    <RenderComponent />
  );
}

function GenerateRoutes({ routes, rootPath }) {
  return (
    <AuthenticatedContext.Consumer>
      {({ isAuthenticated }) => (
        <Routes>
          {routes.map((route) => {
            const path = rootPath ? rootPath + route.path : route.path;

            return (
              <Route
                key={path}
                exact={route.exact}
                path={path}
                element={(
                  <RouteRender
                    isPrivate={route.isPrivate}
                    RenderComponent={route.component}
                    isAuthenticated={isAuthenticated}
                  />
            )}
              />
            );
          })}

          <Route path="*" element={<MissingPage />} />
        </Routes>
      )}
    </AuthenticatedContext.Consumer>
  );
}

RouteRender.defaultProps = {
  isPrivate: false,
};

RouteRender.propTypes = {
  isPrivate: PropTypes.bool,
  isAuthenticated: PropTypes.bool.isRequired,
  RenderComponent: PropTypes.func.isRequired,
};

GenerateRoutes.defaultProps = {
  rootPath: null,
};

GenerateRoutes.propTypes = {
  routes: PropTypes.arrayOf(PropTypes.objectOf(PropTypes.any)).isRequired,
  rootPath: PropTypes.string,
};

export default GenerateRoutes;
