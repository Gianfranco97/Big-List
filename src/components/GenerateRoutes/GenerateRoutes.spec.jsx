import React from 'react';
import { shallow } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import GenerateRoutes from './index';
import { AuthenticatedProvider } from '../AuthenticatedContext';

const demoRoutes = [
  {
    path: '/',
    exact: true,
    component: () => <div />,
    isPrivate: true,
    name: 'Admin',
  },
  {
    path: '/employees',
    isPrivate: true,
    component: () => <div />,
    name: 'Employees',
  },
];

it('The snapshot matches', () => {
  const tree = renderer
    .create(
      <MemoryRouter>
        <AuthenticatedProvider>
          <GenerateRoutes routes={demoRoutes} />
        </AuthenticatedProvider>
      </MemoryRouter>,
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders without crashing', () => {
  shallow(
    <AuthenticatedProvider>
      <GenerateRoutes routes={demoRoutes} />
    </AuthenticatedProvider>,
  );
});
