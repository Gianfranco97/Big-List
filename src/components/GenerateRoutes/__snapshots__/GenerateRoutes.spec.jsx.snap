// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The snapshot matches 1`] = `null`;
